packages/tx/src/legacyTransaction.ts

```diff
@@ -76,6 +77,20 @@ export default class Transaction extends BaseTransaction<Transaction> {
 
     const [nonce, gasPrice, gasLimit, to, value, data, v, r, s] = values
 
+    const integerProps = [nonce, gasPrice, gasLimit, value]
+
+    for (const x of [v, r, s]) {
+      // If transaction is signed, add v, r, and s to values checked for leading zeroes
+      if (x?.length) integerProps.push(x)
+    }
+
+    for (const x of integerProps) {
+      if (x.length > 0 && x[0] === 0x00) {
+        // Throws if integer is encoded with leading zeroes
+        throw new Error('Integer values cannot have leading zeroes')
+      }
+    }
```

Comment:
Some comments here: 😋

1. Again, it would be good if the reasoning is stated here shortly as a one line comment
2. Is there any reason that this is only being done in the legacy transaction?
3. Let's please differentiate the error message with the concrete parameter leading to the error message (even if the code gets slighly more complicated (maybe a small internal function or something).

Reply:
This needs some test cases, minimally:
- Signed tx with focus on the v,r,s values (v is always minimally a good candidate)
- Same for unsigned tx
- Nonce being 0
- Nonce being non-0

Reply:
> Some comments here: yum
> 
> 1. Again, it would be good if the reasoning is stated here shortly as a one line comment

Okay, will add.  It comes down to the [RLP rules](https://eth.wiki/en/fundamentals/rlp) that state that deserialized positive integers with leading zeroes are invalid.

> 2. Is there any reason that this is only being done in the legacy transaction?
User error.  I'll move the check for all the numeric fields in the base transaction to `baseTransaction` and then check transaction type specific fields (e.g. `gasPrice` for legacy transactions, `baseFeePerGas` for EIP1559) to their specific transaction constructors.

> 3. Let's please differentiate the error message with the concrete parameter leading to the error message (even if the code gets slighly more complicated (maybe a small internal function or something).

Will do.

Reply:
> This needs some test cases, minimally:
> 
> * Signed tx with focus on the v,r,s values (v is always minimally a good candidate)

Should be covered by these tests from the Ethereum-Tests repo:
ValidChainID1InvalidV00
TransactionWithSvaluePrefixed00BigInt
TransactionWithRvaluePrefixed00BigInt

> * Same for unsigned tx

Not sure about this one.  Are you just looking for a test where there is no v, r, or s?  I can write one if you don't feel like the existing test suite covers it though I removed the specific logic that only checked an unsigned transaction and replaced with a check that iterates over all of the transaction fields that could be integers (regardless of signed/unsigned status) and checks them for leading zeroes.  Given this reworking of the logic, it seems like the existing tests in the Ethereum/Tests transaction test suite should cover this since they already have scenarios with leading zeroes for all of the integer fields.

> * Nonce being 0

Maybe covered by [this one](https://github.com/ethereum/tests/blob/develop/TransactionTests/ttNonce/TransactionWithEmptyBigInt.json)
> * Nonce being non-0

Should be covered by [this test](https://github.com/ethereum/tests/blob/develop/TransactionTests/ttNonce/TransactionWithLeadingZerosNonce.json)

Reply:
Thought about this a bit longer.

Yes, I guess we also need these tests ourselves. The thing with the external tests is: these kind of tests are very much a black box. One has no real oversight respectively is loosing track what is actually being tested.

But more importantly: these tests at the end only test the general functionality requirements on a tx, for this they are good and useful. They are *not* testing our API though. This is especially crucial for input values. We have so many static constructors. The tx tests only test one of them. If we forget adding a certain type of functionality to one of the constructors not being called, this would already not be caught. Same goes for different input type (`number`, `BN`, `Buffer`,...) e.g.. So we need these additional local tests for this other-perspective testing. Especially on the tx library which is so sensitive.

Reply:
Makese sense.  I'm reworking how I check for leading zeroes and will add appropriate API tests along the way.